### PR TITLE
Make CTR-DRBG use cipher.c when available

### LINKED ISSUE #8054: Make CTR-DRBG use cipher.c when available
Same as #8051 but for `ctr_dbrg.c`.

Note: here ([unlike PEM perhaps](https://github.com/Mbed-TLS/mbedtls/issues/8051#issuecomment-1673044442)) we really don't want to introduce an unconditional dependency on `CIPHER_C`, we'll only depend on `CIPHER_LIGHT` once this has been defined.

### PR DESCRIPTION
**Pull request by valeriosetti**
Resolves #8054

## PR checklist

- [ ] **changelog** not required
- [ ] **backport** not required
- [ ] **tests** not required

### CODE REVIEW

**Comment by valeriosetti** [comment]
The CI only failed in `ABI-API checking` and this is due to the replacement of `aes_ctx` with the new `cipher_ctx` in the `struct mbedtls_ctr_drbg_context` when `CIPHER_C` is defined. Both of these fields are `MBEDTLS_PRIVATE` so no end user is supposed to have direct access to them. At the same time this change is necessary to complete [the original task](https://github.com/Mbed-TLS/mbedtls/issues/8054) so I would say this error is not relevant.
Question to reviewers: wdyt?

**Comment by gilles-peskine-arm** [comment]
Adding and removing `MBEDTLS_PRIVATE` fields is always ok in the development branch. Unfortunately those false positives of the ABI/API checker are not trivial to fix.

**Comment by valeriosetti** [comment]
> Adding and removing MBEDTLS_PRIVATE fields is always ok in the development branch. Unfortunately those false positives of the ABI/API checker are not trivial to fix.

@gilles-peskine-arm wouldn't this create problems with the merge queue? If so, do you have any suggestion on how to overcome this problem?

**Comment by gilles-peskine-arm** [comment]
Only the checks marked “Required” are taken into account for the merge queue. The interface stability tests aren't.

**Review by gabor-mezei-arm — DISMISSED** [review]
LGTM

File: library/ctr_drbg.c [hunk 1]
```diff
@@ -139,9 +159,16 @@ static int block_cipher_df(unsigned char *output,
         return MBEDTLS_ERR_CTR_DRBG_INPUT_TOO_BIG;
     }
 
-    memset(buf, 0, MBEDTLS_CTR_DRBG_MAX_SEED_INPUT +
-           MBEDTLS_CTR_DRBG_BLOCKSIZE + 16);
+    memset(buf, 0, sizeof(buf));
+#if defined(MBEDTLS_CIPHER_C)
+    ret = mbedtls_cipher_setup(&cipher_ctx,
```
Comment (mpg) [inline library/ctr_drbg.c:164]: I think we're supposed to call `init()` before setup. (And also before the first point that might jump to the label where `free()` is called.)

**Review by mpg — CHANGES_REQUESTED** [review]
This is absolutely going in the right direction, but the issue with errors from `setup()` needs addressing.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I question why we're doing this at all. It seems like a step in the wrong direction.

* We have no request to extend CTR_DRBG to other algorithms.
* In Mbed TLS 3.6, we're trying to bypass cipher — see https://github.com/Mbed-TLS/mbedtls/pull/8057. So using cipher more seems counterproductive.
* In Mbed TLS 4.0, we'd like to get rid of cipher. So using cipher more seems counterproductive.

In Mbed TLS 3.6, we may have a goal of making CTR_DRBG work on top of a PSA AES accelerator. It's not yet clear that this is in the scope of 3.6, however. And using cipher.h functions does not seem like a particularly useful way to reach this goal.

**Comment by mpg** [comment]
@gilles-peskine-arm https://github.com/Mbed-TLS/mbedtls/issues/8054#issuecomment-1755027790

**Comment by mpg** [comment]
See https://github.com/Mbed-TLS/mbedtls/issues/8337 regarding clarification of the plan.

But also, there might just be an misunderstanding elsewhere: this PR is only a stepping stone towards #8060 - it's just a way to allow starting the work while #8057 is not completed yet. So the plan is not to introduce more dependencies on `CIPHER_C` (this PR does not add a dependency), just to prepare the move to Cipher-light, taking advantage of the fact it's a subset of Cipher.

**Comment by mpg** [comment]
Ok, following discussion in #8337 there's likely to be a change of plan for CTR-DRBG. Let's put this on hold until we're sure. (I think there's now an agreement between Gilles and I, but I'd like to wait a few more days before jumping to the new plan in order to leave for the consensus to solidify.)

**Pull request closed without merging**